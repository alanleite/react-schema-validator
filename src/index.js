export useStateWithValidator from "./useStateWithValidator"
export schema from "./schema-builder"