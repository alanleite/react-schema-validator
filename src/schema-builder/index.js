import text from './text'

const builder = {
  text 
}

export default builder