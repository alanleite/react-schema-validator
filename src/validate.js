export default (schema, data) => {
    const errors = []
    for (const key in schema) {
      schema[key].validators.forEach(val => {
        const isValid = val.action(data[key])
        if (!isValid) {
          errors.push({
            key: key,
            message: val.message
          })
        }
      })
    }
    return errors.reduce((obj, current) => {
      if(obj[current.key]) {
        obj[current.key].push(current.message)
        return obj
      } else {
        obj[current.key] = [current.message]
        return obj
      }
    }, {})
  }