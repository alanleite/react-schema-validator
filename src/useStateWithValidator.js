import { useState } from 'react'
import validate from './validate'

export default (defaultState, schema) => {
    const [state, setState] = useState(defaultState)
    const setStateValidate = (data) => {
        data._errors = validate(schema, data)
        setState(data)
    }
    return [
        state,
        setStateValidate
    ]
}