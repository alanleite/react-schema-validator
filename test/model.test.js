import React from 'react'
import { render, fireEvent } from 'react-testing-library'
import { useStateWithValidator, schema } from '../src'

const Page = () => {
    const defaultData = { name: 'Jonhy' }
    const definitions = { name: schema.text().max(2) }
    const [state, setState] = useStateWithValidator(defaultData, definitions)
    
    const changeName = () => {
        setState({
            name: 'Teste'
        })
    }

    return (
        <div>
            {state.name}
            {JSON.stringify(state._errors)}
            <button onClick={changeName}>Change</button>
        </div>
    )
}

describe('Button component', () => {
    it('deve conseguir renderizar [snapshot]', () => {
      const { container, getByText } = render(<Page />)
      expect(container.firstChild).toMatchSnapshot()
      fireEvent.click(getByText('Change'))
      expect(container.firstChild).toMatchSnapshot()
    })
  })