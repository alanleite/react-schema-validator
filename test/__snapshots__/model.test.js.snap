// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component deve conseguir renderizar [snapshot] 1`] = `
<div>
  Jonhy
  <button>
    Change
  </button>
</div>
`;

exports[`Button component deve conseguir renderizar [snapshot] 2`] = `
<div>
  Teste
  {"name":["Quantidade máxima de caracteres é 2"]}
  <button>
    Change
  </button>
</div>
`;
